sarc/jobs/series.py

```diff
@@ -479,3 +479,91 @@ def _compute_cost_and_wastes(data, device):
     )
 
     return data
+
+
+def compute_time_frames(
+    jobs,
+    columns: list[str] | None = None,
+    start: datetime | None = None,
+    end: datetime | None = None,
+    frame_size: timedelta = timedelta(days=7),
+):
+    """Slice jobs into time frames and adjust columns to fit the time frames.
+
+    Jobs that start before `start` or ends after `end` will have their running
+    time clipped to fitting within the interval (`start`, `end`).
+
+    Jobs spanning multiple time frames will have their running time sliced
+    according to the time frames.
+
+    The resulting DataFrame will have the additional columns 'duration' and 'timestamp'
+    which represent the duration of a job within a time frame and the start of the time frame.
+
+    Parameters
+    ----------
+    jobs: DataFrame
+        Pandas DataFrame containing jobs data. Typically generated with `load_job_series`.
+        Must contain columns `start_time` and `end_time`.
+    columns: list of str
+        Columns to adjust based on time frames.
+    start: datetime, optional
+        Start of the time frame. If None, use the first job start time.
+    end: datetime, optional
+        End of the time frame. If None, use the last job end time.
+    frame_size: timedelta, optional
+        Size of the time frames used to compute histograms. Default to 7 days.
+
+    Examples
+    --------
+    >>> data = pd.DataFrame(
+        [
+            [datetime(2023, 3, 5), datetime(2023, 3, 6), "a", "A", 10],
+            [datetime(2023, 3, 6), datetime(2023, 3, 9), "a", "B", 10],
+            [datetime(2023, 3, 6), datetime(2023, 3, 7), "b", "B", 20],
+            [datetime(2023, 3, 6), datetime(2023, 3, 8), "b", "B", 20],
+        ],
+        columns=["start", "end", "user", "cluster", 'cost'],
+    )
+    >>> compute_time_frames(data, columns=['cost'], frame_size=timedelta(days=2))
+           start        end user cluster       cost  duration  timestamp
+    0 2023-03-05 2023-03-06    a       A  10.000000   86400.0 2023-03-05
+    1 2023-03-06 2023-03-09    a       B   3.333333   86400.0 2023-03-05
+    2 2023-03-06 2023-03-07    b       B  20.000000   86400.0 2023-03-05
+    3 2023-03-06 2023-03-08    b       B  10.000000   86400.0 2023-03-05
+    1 2023-03-06 2023-03-09    a       B   6.666667  172800.0 2023-03-07
+    3 2023-03-06 2023-03-08    b       B  10.000000   86400.0 2023-03-07
+    """
+    col_start = "start_time"
+    col_end = "end_time"
+
+    if columns is None:
+        columns = []
+
+    if start is None:
+        start = jobs[col_start].min()
+
+    if end is None:
+        end = jobs[col_end].max()
+
+    data_frames = []
+
+    total_durations = (jobs[col_end] - jobs[col_start]).dt.total_seconds()
+    for frame_start in pandas.date_range(start, end, freq=f"{frame_size.days}D"):
```

Comment:
If frame_size is less than one day, this makes freq="0d" which raises a ZeroDivisionError. date_range does accept a timedelta directly, though, so this works just fine:

```suggestion
    for frame_start in pandas.date_range(start, end, freq=frame_size):
```